Assignments/A3/Distro/QLearn.c

```diff
@@ -416,3 +438,74 @@ int pos_to_index(int pos[2], int size_X) { return pos[0] + pos[1] * size_X; }
 int is_pos_valid(int pos[2], int size_X, int size_Y) {
   return 0 <= pos[0] && pos[0] < size_X && 0 <= pos[1] && pos[1] < size_Y;
 }
+
+// Feature: Returns the avg distance to the cats from the mouse between [-1, 1]
+double avg_cat_feat(double gr[max_graph_size][4], double features[25],
+                    int mouse_pos[1][2], int cats[5][2], int size_X,
+                    int graph_size) {
+  // -1 is terrible, aka cats are right on top
+  // 1 is great, aka cats are across the map from the mouse
+  double max_dist =
+      sqrt(pow((double)(size_X), 2) + pow((double)(graph_size / size_X), 2));
+  double average_cat_distance = 0;
+  int num_cat;
+  for (num_cat = 0; cats[num_cat][0] != -1; ++num_cat) {
+    const double kCatDistance =
+        sqrt(pow((double)(mouse_pos[0][0] - cats[num_cat][0]), 2) +
+             pow((double)(mouse_pos[0][1] - cats[num_cat][1]), 2));
+    average_cat_distance += kCatDistance;
+  }
+  // Get avg distance and normalize on max distance to [0,1]
+  average_cat_distance /= num_cat * max_dist;
+  // manipulate range to [-0.5, 0.5] *2 = [-1, 1]
+  return (average_cat_distance - 0.5) * 2;
+}
+
+/* Feature: Returns the distance to the closest agent in agents from the mouse
+ between [-1, 1] Possible agents: Cat, Cheese*/
+double closest_dist(double gr[max_graph_size][4], double features[25],
+                    int mouse_pos[1][2], int agents[5][2], int size_X,
+                    int graph_size) {
+  // -1 is terrible, aka cats are right on top
+  // 1 is great, aka cats are across the map from the mouse
+  double max_dist =
+      sqrt(pow((double)(size_X), 2) + pow((double)(graph_size / size_X), 2));
+  double closest_dist = max_dist;
+
+  for (int num_agent = 0; agents[num_agent][0] != -1; ++num_agent) {
+    double dist =
+        sqrt(pow((double)(mouse_pos[0][0] - agents[num_agent][0]), 2) +
+             pow((double)(mouse_pos[0][1] - agents[num_agent][1]), 2));
+    closest_dist = fmin(dist, closest_dist);
+  }
+  closest_dist /= max_dist;
+  return (closest_dist - 0.5) * 2;
+}
+
+// Check if currenty at dead end
+bool dead_end(double gr[max_graph_size][4], int mouse_pos[1][2], int size_X) {
+  int walls = 0;
+  for (int i = 0; i < 4; ++i) {
+    walls += gr[pos_to_index(mouse_pos[0], size_X)][i];
+  }
+  return (walls > 2);
+}
+
+// Define a function that calculates the angle between three points
+double angle(int mouse_pos[1][2], int cats[5][2], int cheese[2]) {
+  int num_c = 0, c_x = 0, c_y = 0;
+  for (num_c = 0;cats[num_c][0] != -1; ++num_c) {
+    c_x += cats[num_c][0];
+    c_y += cats[num_c][1];
+  }
+
+  // Calculate the lengths of the sides
+  double mx = (c_x / num_c) - mouse_pos[0][0];
+  double my = (c_y / num_c) - mouse_pos[0][1];
```

Comment:
`c_x / num_c` uses integer division, so should probably be:
`(double) c_x / (double) num_c`